src/segwit.rs

```diff
@@ -0,0 +1,266 @@
+//! [BIP-350]: <https://github.com/bitcoin/bips/blob/master/bip-0350.mediawiki>
+//! [`bip_173_test_vectors.rs`]: <https://github.com/rust-bitcoin/rust-bech32/blob/master/tests/bip_173_test_vectors.rs>
+//! [`bip_350_test_vectors.rs`]: <https://github.com/rust-bitcoin/rust-bech32/blob/master/tests/bip_350_test_vectors.rs>
+
+#[cfg(all(feature = "alloc", not(feature = "std"), not(test)))]
+use alloc::{string::String, vec::Vec};
+use core::fmt;
+
+#[cfg(feature = "alloc")]
+use crate::primitives::decode::{SegwitHrpstring, SegwitHrpstringError};
+use crate::primitives::gf32::Fe32;
+use crate::primitives::hrp::Hrp;
+use crate::primitives::iter::{ByteIterExt, Fe32IterExt};
+use crate::primitives::segwit::{self, InvalidWitnessVersionError, WitnessLengthError};
+pub use crate::primitives::segwit::{VERSION_0, VERSION_1};
+use crate::primitives::{Bech32, Bech32m};
+use crate::write_err;
+
+/// Decodes a segwit address.
+///
+/// # Examples
+///
+/// ```
+/// use bech32::segwit;
+/// let address = "bc1py3m7vwnghyne9gnvcjw82j7gqt2rafgdmlmwmqnn3hvcmdm09rjqcgrtxs";
+/// let (_hrp, _witness_version, _witness_program) = segwit::decode(address).expect("failed to decode address");
+/// ```
+#[cfg(feature = "alloc")]
+pub fn decode(s: &str) -> Result<(Hrp, Fe32, Vec<u8>), SegwitHrpstringError> {
+    let segwit = SegwitHrpstring::new(s)?;
+    Ok((segwit.hrp(), segwit.witness_version(), segwit.byte_iter().collect::<Vec<u8>>()))
+}
+
+/// Encodes a segwit address.
+///
+/// Does validity checks on the `witness_version` and length checks on the `witness_program`.
+///
+/// As specified by [`BIP-350`] we use the [`Bech32m`] checksum algorithm for witness versions 1 and
+/// above, and for witness version 0 we use the original ([`BIP-173`]) [`Bech32`] checksum
+/// algorithm.
+///
+/// See also [`encode_v0`] or [`encode_v1`].
+///
+/// [`Bech32`]: crate::primitives::Bech32
+/// [`Bech32m`]: crate::primitives::Bech32m
+/// [BIP-173]: <https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki>
+/// [BIP-350]: <https://github.com/bitcoin/bips/blob/master/bip-0350.mediawiki>
+#[cfg(feature = "alloc")]
+pub fn encode(
+    hrp: &Hrp,
+    witness_version: Fe32,
+    witness_program: &[u8],
+) -> Result<String, EncodeError> {
+    segwit::validate_witness_version(witness_version)?;
+    segwit::validate_witness_program_length(witness_program.len(), witness_version)?;
+
+    let mut buf = String::new();
+    encode_to_fmt_unchecked(&mut buf, hrp, witness_version, witness_program)?;
+    Ok(buf)
+}
+
+/// Encodes a segwit version 0 address.
+#[cfg(feature = "alloc")]
+pub fn encode_v0(hrp: &Hrp, witness_program: &[u8]) -> Result<String, EncodeError> {
+    encode(hrp, segwit::VERSION_0, witness_program)
+}
+
+/// Encodes a segwit version 1 address.
+#[cfg(feature = "alloc")]
+pub fn encode_v1(hrp: &Hrp, witness_program: &[u8]) -> Result<String, EncodeError> {
+    encode(hrp, segwit::VERSION_1, witness_program)
```

Comment:
No need for `segwit::` scoping here.

Reply:
I umm'ed and ahhh'ed about that line, whether the "segwit" added anything, will remove.